core/src/main/java/com/yahoo/imapnio/async/request/AbstractSearchCommand.java

```diff
@@ -0,0 +1,191 @@
+import javax.mail.search.SearchTerm;
+
+import com.sun.mail.iap.Argument;
+import com.sun.mail.iap.ProtocolException;
+import com.sun.mail.imap.protocol.SearchSequence;
+import com.yahoo.imapnio.async.data.Capability;
+import com.yahoo.imapnio.async.data.MessageNumberSet;
+import com.yahoo.imapnio.async.exception.ImapAsyncClientException;
+import com.yahoo.imapnio.async.exception.ImapAsyncClientException.FailureType;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+/**
+ * This class defines IMAP search command request from client.
+ *
+ * <pre>
+ * search         = "SEARCH" [SP "CHARSET" SP astring] 1*(SP search-key)
+ *                   ; CHARSET argument to MUST be registered with IANA
+ *
+ * search-key     = "ALL" / "ANSWERED" / "BCC" SP astring /
+ *                  "BEFORE" SP date / "BODY" SP astring /
+ *                  "CC" SP astring / "DELETED" / "FLAGGED" /
+ *                  "FROM" SP astring / "KEYWORD" SP flag-keyword /
+ *                  "NEW" / "OLD" / "ON" SP date / "RECENT" / "SEEN" /
+ *                  "SINCE" SP date / "SUBJECT" SP astring /
+ *                  "TEXT" SP astring / "TO" SP astring /
+ *                  "UNANSWERED" / "UNDELETED" / "UNFLAGGED" /
+ *                  "UNKEYWORD" SP flag-keyword / "UNSEEN" /
+ *                    ; Above this line were in [IMAP2]
+ *                  "DRAFT" / "HEADER" SP header-fld-name SP astring /
+ *                  "LARGER" SP number / "NOT" SP search-key /
+ *                  "OR" SP search-key SP search-key /
+ *                  "SENTBEFORE" SP date / "SENTON" SP date /
+ *                  "SENTSINCE" SP date / "SMALLER" SP number /
+ *                  "UID" SP sequence-set / "UNDRAFT" / sequence-set /
+ *                  "(" search-key *(SP search-key) ")"
+ * </pre>
+ */
+public abstract class AbstractSearchCommand extends ImapRequestAdapter {
+
+    /** Character set literal. */
+    private static final String CHARSET = "CHARSET";
+
+    /** Charset literal following by a space. */
+    private static final byte[] CHARSET_B = CHARSET.getBytes(StandardCharsets.US_ASCII);
+
+    /** SEARCH . */
+    private static final String SEARCH = "SEARCH";
+
+    /** SEARCH in byte array. */
+    private static final byte[] SEARCH_B = SEARCH.getBytes(StandardCharsets.US_ASCII);
+
+    /** UID SEARCH. */
+    private static final String UID_SEARCH = "UID SEARCH";
+
+    /** UID SEARCH in byte array. */
+    private static final byte[] UID_SEARCH_B = UID_SEARCH.getBytes(StandardCharsets.US_ASCII);
+
+    /** Flag whether adding UID before search. */
+    private boolean isUid;
+
+    /** Message numbers in string type, specified based on RFC3501 sequence-set syntax. */
+    private String msgNumbers;
+
+    /** The search expression. */
+    private Argument searchExpr;
+
+    /** Character set. */
+    private String charset;
+
+    /** flag whether server allows LITERAL+. */
+    private boolean isLiteralPlusEnabled;
```

Comment:
Hmm. Fyi, this is something related to data streaming. Should this be included in request?